### PR TITLE
feat(broadcast): BroadcastEvent schema + classify_op + redact_payload (G6.1-T2)

### LINKED ISSUE #308: Task: G6.1-T2 BroadcastEvent schema + classify_op + redact_payload (conservative PII defaults per decision #3)
Parent initiative: #228
Depends on: G6.1-T1.

## Target

`BroadcastEvent` Pydantic schema + sensitivity classifier per the v0.2 PII defaults (decision #3 in [v0.2-decisions.md](docs/planning/v0.2-decisions.md)):

- `backend/src/meho_backplane/broadcast/events.py` — `BroadcastEvent` model + `classify_op` + `redact_payload`.
- `backend/tests/test_broadcast_events.py` — unit tests for the classifier.

## Context

T2 ships the **event contract** + **PII discipline** that every downstream broadcast consumer relies on. The classifier locks how sensitive ops are redacted before they reach the SSE feed or Slack mirror.

**Event schema:**

```python
class BroadcastEvent(BaseModel):
    model_config = ConfigDict(frozen=True)

    event_id: UUID
    ts: datetime
    tenant_id: UUID
    principal_sub: str
    principal_name: str | None
    target_name: str | None
    op_id: str
    op_class: str               # "read" | "write" | "credential_read" | "audit_query" | "other"
    result_status: str          # "ok" | "error" | "denied"
    audit_id: UUID              # FK to audit_log
    payload: dict[str, Any]     # redacted per classifier
```

**Classifier rules (decision #3 — conservative):**

```python
_CREDENTIAL_READ_OPS = {"vault.kv.read", "vault.kv.list"}  # extensible registry

def classify_op(op_id: str) -> str:
    if op_id in _CREDENTIAL_READ_OPS:
        return "credential_read"
    if op_id.startswith("audit."):
        return "audit_query"
    if op_id.endswith((".create", ".update", ".delete", ".patch")):
        return "write"
    if op_id.endswith((".list", ".info", ".get", ".about", ".ls")):
        return "read"
    return "other"


def redact_payload(op_class: str, raw_params: dict, result_status: str) -> dict:
    if op_class == "credential_read":
        return {"op_class": op_class, "result_status": result_status}
    if op_class == "audit_query":
        return {
            "op_class": op_class,
            "result_status": result_status,
            "row_count": _maybe_row_count(raw_params),
        }
    # read / write / other: full detail
    return {
        "op_class": op_class,
        "params": raw_params,
        "result_status": result_status,
    }
```

**Why a classifier (not a per-op annotation)?** Two reasons:
1. Sensitivity is mostly op-class-shaped; per-op fine-tuning would explode the contract surface.
2. The classifier is one auditable function — easier to review for policy compliance than scattered annotations.

Per-op opt-in (e.g., "broadcast vault.kv.read in full detail because we trust this tenant") is a G6.3 feature; T2 ships the conservative default.

## Current state

After T1: Valkey client wired. No event schema, no classifier.

## Desired state

`events.py` per the sketch above. Tested against representative op_ids.

## Acceptance criteria

- [ ] `BroadcastEvent` schema defined; frozen; all required fields populated by callers.
- [ ] `classify_op("vault.kv.read")` → `"credential_read"`.
- [ ] `classify_op("audit.query")` → `"audit_query"`.
- [ ] `classify_op("vsphere.vm.list")` → `"read"`.
- [ ] `classify_op("vsphere.vm.create")` → `"write"`.
- [ ] `classify_op("some.unknown.op")` → `"other"`.
- [ ] `redact_payload("credential_read", {"path": "secret/foo"}, "ok")` → `{op_class, result_status}` (path NOT present).
- [ ] `redact_payload("audit_query", {"filter": "X"}, "ok")` → `{op_class, result_status, row_count}` (filter NOT present).
- [ ] `redact_payload("read", {"folder": "X"}, "ok")` → `{op_class, params={"folder": "X"}, result_status}` (folder IS present).
- [ ] Unit tests cover each branch + edge cases (empty params, params with nested objects).
- [ ] CI green.
- [ ] [#228](https://github.com/evoila/meho/issues/228) body checklist line for T2 is ticked.

## Out of scope

- Per-op opt-in/opt-out (G6.3).
- Per-tenant convention overrides for classification.
- Operator-side redaction toggle.
- ML-based PII detection.

## References

- Parent Initiative: [#228](https://github.com/evoila/meho/issues/228).
- Decision #3 in [docs/planning/v0.2-decisions.md](docs/planning/v0.2-decisions.md).
- Consumer-needs.md §G6 L368.

### PR DESCRIPTION
**Pull request by zdamir**
## Summary

Closes #308 — T2 lands the event contract every downstream broadcast consumer (T3 publish hook, T4 SSE, T6 MCP resource) reads back from the Valkey stream. Locks [decision #3](../blob/main/docs/planning/v0.2-decisions.md)'s conservative PII defaults so the publish hook in T3 inherits a redaction policy already cast in stone — no per-tenant or per-op opt-in surface here (that's G6.3).

- **`broadcast/events.py` (new)** — `BroadcastEvent` Pydantic v2 model with `ConfigDict(frozen=True)`. Frozen so a consumer mid-pipeline can't mutate fields the next stage assumed immutable. Mirrors the chassis `Operator` pattern.
- **`classify_op(op_id)`** — five-class derivation: `credential_read` (exact allowlist `vault.kv.read` + `vault.kv.list`), `audit_query` (`audit.*` prefix), `write` (`.create` / `.update` / `.delete` / `.patch` suffixes), `read` (`.list` / `.info` / `.get` / `.about` / `.ls` suffixes), `other` (default). Match order is policy-significant — credential allowlist runs before the suffix check so a future `vault.kv.audit-list` could opt out by being absent from the allowlist.
- **`redact_payload(op_class, raw_params, result_status)`** — per-class redaction. `credential_read` collapses to `{op_class, result_status}` (no path, no key, no value); `audit_query` adds `{row_count}` extracted via `_maybe_row_count` with int-coercion + fail-closed-to-None on a malformed JSON round-trip; `read` / `write` / `other` keep full params verbatim (no defensive copies — nested objects pass through).
- **Tests:** 41 cases across schema (frozen, required, optional, omittable, JSON round-trip), `classify_op` (per-class parametric + precedence pins for credential allowlist beats `.list` suffix and `audit.` prefix beats both `.list` and `.query` suffixes), `redact_payload` (per-class shape + absent-key invariants for credential `path` / `key` and audit `filter`, plus stringified `row_count` coercion + nested-params passthrough).

## Test plan

- [x] `ruff check src tests` — clean
- [x] `ruff format --check src tests` — clean (102 files)
- [x] `mypy src` strict — 53 source files, no issues
- [x] `pytest tests/test_broadcast_events.py` — **41 passed**
- [x] `pytest --ignore=tests/integration` — **514 passed** (+41 vs main), 1 skipped, 1 deselected
- [x] AC #1 — `BroadcastEvent` defined, frozen, all required fields enforced → `TestBroadcastEventSchema`
- [x] AC #2-5 — `classify_op` mapping for every named op-id → `TestClassifyOp` parametric cases
- [x] AC #6 — `redact_payload("credential_read", ...)` drops path/key → `test_credential_read_drops_path_and_key`
- [x] AC #7 — `redact_payload("audit_query", ...)` drops filter → `test_audit_query_drops_filter`
- [x] AC #8 — `redact_payload("read", ...)` keeps params → `test_read_keeps_full_params`
- [x] AC #9 — unit tests cover each branch + edge cases (empty params, nested objects, stringified row_count, omittable optional fields, default_factory exercised) → all in `TestRedactPayload` / `TestBroadcastEventSchema`

## Out of scope

- Per-op opt-in/opt-out — G6.3 surface (separate Initiative, post-G6.1).
- Per-tenant convention overrides for classification.
- ML-based PII detection.
- T3 publish-on-write hook — separate task #309, builds on this schema.
- `Mapping[str, Any]` + `MappingProxyType` migration for genuine payload immutability — heavy-lift; tracked for v0.2.next if the broadcast pipeline grows transformation stages.

## Post-merge follow-up

- Tick T2 line on Initiative #228 body checklist.

---

## Follow-up (auto-implement-issue, iteration 2, 2026-05-13)

Addressed all three review findings from [/auto-review-pr iter-1](https://github.com/evoila/meho/pull/346#issuecomment-3525914732):

- **B1** `e7e8fc2` — `principal_name` / `target_name` now carry `= None` defaults so T3's publisher can omit them. Strengthened `test_optional_fields_omittable_at_construction` to pin the omission path (the prior `test_optional_fields_accept_none` only covered explicit-`None`).
- **M1** `e7e8fc2` — Class docstring tightened: `frozen=True` now correctly described as faux-immutability (blocks attribute reassignment but NOT nested in-place mutation). Explicit statement that PII enforcement is upstream of the model (`redact_payload()` before construction). `MappingProxyType` migration noted as tracked out-of-scope.
- **M2** `e7e8fc2` — `test_payload_default_is_empty_dict` rewritten to actually exercise `Field(default_factory=dict)` via `model_dump().pop('payload')` + reconstruct; `assert == {}` added so the shared-mutable-default footgun would fail the test if reintroduced.

Closes #308

## Summary by CodeRabbit

* **New Features**
  * Added a frozen BroadcastEvent schema plus classification and payload redaction utilities to record audited operations with conservative PII redaction rules.

* **Tests**
  * Added comprehensive tests for the event schema, classification precedence, immutability, JSON round-trip, and per-class redaction/coercion behavior.

* **Documentation**
  * Updated backend docs to describe the broadcast event schema and the data protection/classification behavior.

[Review Change Stack](https://app.coderabbit.ai/change-stack/evoila/meho/pull/346)

### CODE REVIEW

**Comment by zdamir** [comment]
@coderabbitai review

File: backend/src/meho_backplane/broadcast/events.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    """
+
+    model_config = ConfigDict(frozen=True)
+
+    event_id: UUID
+    ts: datetime
+    tenant_id: UUID
+    principal_sub: str
+    principal_name: str | None
+    target_name: str | None
```
Comment (zdamir) [inline backend/src/meho_backplane/broadcast/events.py]: **[🛑 B1]** — Optional identity fields are required at construction (contract bug; also flagged by CodeRabbit)

**What:** Lines 138-139:
```python
principal_name: str | None
target_name: str | None
```
The fields are annotated nullable but **not** given a `None` default, so Pydantic v2 marks them as required. Verified locally:
```
ValidationError: 2 validation errors for BroadcastEvent
principal_name
  Field required [type=missing, …]
target_name
  Field required [type=missing, …]
```

**Why it matters:** T3's publish hook (#309) will mint a `BroadcastEvent` per audit-log write. `principal_name` is described in the issue spec and module docstring as "best-effort from the JWT's `name` claim" — JWTs without a `name` claim are routine, and the publisher should be able to omit the field rather than be forced to pass `principal_name=None` explicitly. `target_name` has the same shape (unknown until the connector resolves it). The PR's own `test_optional_fields_accept_none` only proves they accept `None` when explicitly passed, not that they're omittable — the contract gap silently passes today and crashes at T3 wire-up.

**Required change:** Add the `None` defaults (CodeRabbit's suggestion verbatim):

```diff
-    principal_name: str | None
-    target_name: str | None
+    principal_name: str | None = None
+    target_name: str | None = None
```

Then strengthen `test_optional_fields_accept_none` to also cover the omission path:

```python
def test_optional_fields_omittable_at_construction(self) -> None:
    """Pydantic must accept events without principal_name / target_name."""
    event = BroadcastEvent(
        event_id=_EVENT_ID, ts=_TS, tenant_id=_TENANT_A,
        principal_sub="op-test", op_id="vsphere.vm.list",
        op_class="read", result_status="ok", audit_id=_AUDIT_ID,
    )
    assert event.principal_name is None
    assert event.target_name is None
```

**Source:** Pydantic v2 — fields without a default value are required regardless of `Optional` / `T | None` annotation. https://docs.pydantic.dev/2.13/concepts/fields/#optional-and-nullable-fields

File: backend/src/meho_backplane/broadcast/events.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    :class:`~meho_backplane.auth.operator.Operator` model documents.
+
+    ``payload`` is **always** the redacted view per
+    :func:`redact_payload` — callers MUST NOT pass raw params here.
+    The class can't enforce this from the type system alone (any
+    ``dict[str, Any]`` satisfies the annotation), so the contract is
+    documented in the publisher's docstring (T3) and verified by
+    integration tests that read back from the stream and assert
+    forbidden keys are absent.
+    """
```
Comment (zdamir) [inline backend/src/meho_backplane/broadcast/events.py:139]: **[⚠️ M1]** — Docstring overstates `frozen=True` immutability for nested mutable fields (also flagged by CodeRabbit)

**What:** Lines 116-121:
> "The model is **frozen** (`ConfigDict(frozen=True)`) so once constructed it cannot be mutated. Downstream consumers may pass events through transformation pipelines without risk of an intermediate stage rewriting fields the next stage assumed immutable."

Pydantic v2's `frozen=True` is faux-immutability — it overrides `__setattr__` to block attribute reassignment (`event.payload = new_dict` raises) but does **NOT** prevent in-place mutation of nested mutable values (`event.payload["k"] = "v"` succeeds silently). The docstring's "no risk of intermediate-stage rewriting" claim is therefore false for `payload: dict[str, Any]` — a transformation stage CAN mutate the dict in place.

**Why it matters:** T3 / T4 / T6 implementers reading this docstring will inherit a misleading promise. The chassis `Operator` model the docstring cites carries primitive-typed fields, so the analogy doesn't hold for `BroadcastEvent.payload`. Locking the wrong contract in writing means a future "let's transform the payload mid-pipeline" change won't trip any review.

**Required change:** Tighten the docstring to match the actual contract — frozen blocks reassignment, not nested mutation, and the redaction guarantee comes from the publisher calling `redact_payload()` before construction, not from the model. Minimal diff:

```diff
-    The model is **frozen** (``ConfigDict(frozen=True)``) so once
-    constructed it cannot be mutated. Downstream consumers may pass
-    events through transformation pipelines without risk of an
-    intermediate stage rewriting fields the next stage assumed
-    immutable. Same rationale the chassis
-    :class:`~meho_backplane.auth.operator.Operator` model documents.
+    The model is **frozen** (``ConfigDict(frozen=True)``) — attribute
+    reassignment (``event.payload = new_dict``) raises, mirroring the
+    chassis :class:`~meho_backplane.auth.operator.Operator` pattern.
+    ``frozen=True`` is faux-immutability in pydantic v2: nested mutable
+    values like the ``payload`` dict CAN still be mutated in place
+    (``event.payload["k"] = "v"`` succeeds). Downstream consumers MUST
+    NOT mutate ``payload`` between construction and publish; the
+    redaction contract is enforced upstream by the publisher calling
+    :func:`redact_payload` before construction, not by the model itself.
+    A future tightening to ``Mapping`` + ``MappingProxyType`` is
+    tracked but out of scope for T2.
```

Picking the docstring fix over CodeRabbit's `MappingProxyType` suggestion deliberately — the practical mutation window is "between `BroadcastEvent(...)` and `XADD`", which is microseconds in the T3 publisher and zero for T4/T6 consumers that deserialise from JSON (fresh dicts every time). The heavy-lift `Mapping` migration is worth tracking as an adjacent improvement but not warranted for T2's blast radius.

**Source:** Pydantic v2 frozen-models docs — https://docs.pydantic.dev/2.13/concepts/models/#faux-immutability (link confirms the "blocks reassignment, doesn't prevent in-place mutation" behaviour CodeRabbit's web query also surfaced).

File: backend/tests/test_broadcast_events.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+    def test_payload_default_is_empty_dict(self) -> None:
+        """Default payload lands at ``{}``, not a shared mutable instance."""
+        a = _make_event()
+        b = _make_event()
+        # Mutating the dict on ``a`` after construction would mutate ``b`` too
+        # if the field's default_factory were a shared instance; the frozen
+        # model itself blocks that path, but verifying object identity
+        # documents the contract.
+        assert a.payload is not b.payload
+
```
Comment (zdamir) [inline backend/tests/test_broadcast_events.py:142]: **[⚠️ M2]** — `test_payload_default_is_empty_dict` doesn't exercise the default_factory (also flagged by CodeRabbit)

**What:** Lines 95-104:
```python
def test_payload_default_is_empty_dict(self) -> None:
    """Default payload lands at ``{}``, not a shared mutable instance."""
    a = _make_event()
    b = _make_event()
    # … `assert a.payload is not b.payload`
```
`_make_event()` always supplies a `payload` kwarg (defaulting to a fresh dict literal in the fixture body), so the `default_factory=dict` code path on `BroadcastEvent.payload` is never traversed. The `a.payload is not b.payload` assertion happens to pass — but because `_make_event` constructs a new dict literal per call, not because the model's `default_factory` is being verified. The test's docstring is misleading.

**Why it matters:** The default_factory exists specifically to prevent the classic "mutable default argument shared across instances" footgun. If a future refactor accidentally swaps `default_factory=dict` for `default={}` (the literal), every event would share one dict, T3 publishers would silently bleed payload data across audited operations, and this test would still pass — defeating its stated purpose.

**Required change:** Construct events that actually omit `payload` so the default_factory runs. CodeRabbit's suggested fix is right; tightening the assertion to also verify `a.payload == {}` makes the test's claim explicit:

```diff
     def test_payload_default_is_empty_dict(self) -> None:
         """Default payload lands at ``{}``, not a shared mutable instance."""
-        a = _make_event()
-        b = _make_event()
+        base = _make_event().model_dump()
+        base.pop("payload", None)
+        a = BroadcastEvent(**base)
+        b = BroadcastEvent(**base)
+        assert a.payload == {}
+        assert b.payload == {}
         # Mutating the dict on ``a`` after construction would mutate ``b`` too
         # if the field's default_factory were a shared instance; the frozen
         # model itself blocks that path, but verifying object identity
         # documents the contract.
         assert a.payload is not b.payload
```

**Source:** Pydantic v2 `Field(default_factory=...)` — https://docs.pydantic.dev/2.13/concepts/fields/#default-values

**Review by zdamir — COMMENTED** [review]
> **Note:** Posting as `--comment` because GitHub blocks formal `REQUEST_CHANGES` on one's own PR. Formal verdict below is **REQUEST_CHANGES**; the orchestrator reads the result file, not this badge. One blocker (B1) + two majors (M1, M2) carry the punch list for `/auto-implement-issue --continue-pr`.

## /auto-review-pr summary (iteration 1)




**Anchor task:** #308 — G6.1-T2 BroadcastEvent schema + classify_op + redact_payload (conservative PII defaults per decision #3)
**Surface area:** backend (Python)
**Best-practices packs audited:** python (pydantic v2, frozen-model semantics, test honesty)
**Verdict:** REQUEST_CHANGES
**Head SHA reviewed:** `4b26edfc4c01f843807f66b709e1b8452eb54d3e`
**AI reviewer (CodeRabbit):** received in ~9 min; 3 actionable findings (all confirmed and adopted as B1 / M1 / M2 below).

### Acceptance criteria
- [x] AC #1 — `BroadcastEvent` defined, frozen, all required fields enforced — `TestBroadcastEventSchema` (5 tests). **Caveat:** the "frozen" claim in the model docstring overstates what `ConfigDict(frozen=True)` actually guarantees for nested mutable fields — see M1.
- [x] AC #2-5 — `classify_op` mapping for every named op-id — `TestClassifyOp` parametric cases including the credential-allowlist-beats-`.list`-suffix and `audit.`-prefix-beats-`.query`-suffix precedence pins.
- [x] AC #6 — `redact_payload("credential_read", ...)` drops path/key — `test_credential_read_drops_path_and_key`.
- [x] AC #7 — `redact_payload("audit_query", ...)` drops filter — `test_audit_query_drops_filter`.
- [x] AC #8 — `redact_payload("read", ...)` keeps params — `test_read_keeps_full_params`.
- [⚠️] AC #9 — unit tests cover each branch + edge cases — mostly yes, but **`test_payload_default_is_empty_dict` doesn't actually exercise the default_factory** (see M2). Other edge cases (empty params, nested objects, stringified row_count) ARE covered.
- [x] AC #10 — CI green — every code gate green except CodeRabbit (PENDING is just the third-party status check still finalising; the review itself landed with findings). DCO, SonarCloud Quality Gate, Python, Semgrep SAST/OSS, Go, Helm, image build, license checks — all SUCCESS.
- [ ] AC #11 — Initiative #228 body checklist line for T2 ticked — post-merge action.

**Contract bug (B1):** `principal_name: str | None` and `target_name: str | None` are required at construction despite the nullable annotation. Verified locally:
```
ValidationError: 2 validation errors for BroadcastEvent
principal_name  Field required [type=missing, …]
target_name     Field required [type=missing, …]
```
T3's publisher will hit this on the first wire-up. AC #1 says "all required fields populated by callers" — these fields are described in the spec as best-effort/optional, so requiring them at construction violates the contract the AC implies.

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #308 (T2) | open | `Closes #308` | ✅ | ⚠️ partial (B1 contract bug) | close on merge after B1 fix |
| #228 (Initiative G6.1) | open | parent of #308, not linked from PR | — | ⚠️ partial (T2 of 6) | tick T2 checkbox post-merge once #308 closes |
| #309 (T3 publish-on-write) | open | downstream consumer of `BroadcastEvent` | — | — | will need an integration test pinning the redaction contract end-to-end |

### Predicted vs. actual diff
- ✅ `backend/src/meho_backplane/broadcast/__init__.py` — re-exports the new public surface.
- ⚠️ `backend/src/meho_backplane/broadcast/events.py` — schema + classifier + redactor are correct in shape, but the model carries the B1 contract bug + the M1 docstring overpromise.
- ⚠️ `backend/tests/test_broadcast_events.py` — comprehensive coverage, but M2 marks one test that doesn't pin what its docstring claims.
- ✅ `docs/codebase/backend.md` — T2 paragraph added at the right anchor; no .NET analogies; the T1 paragraph correctly bumped to reference T3-T6 (T2 now done).

### Integration verification
- [x] `ruff check src tests` — clean
- [x] `ruff format --check src tests` — clean (102 files)
- [x] `mypy src` strict — 53 source files, no issues
- [x] `pytest tests/test_broadcast_events.py` — 40 passed, 7.17 s
- [x] Inline reproduction of B1: `uv run python -c "BroadcastEvent(...with principal_name/target_name omitted...)"` → `ValidationError` (proof posted in B1's inline comment).
- [⚠️ not run] integration test pinning the redaction contract end-to-end against a real Valkey stream — out of scope for T2; T3 will add it.
- [⚠️ not run] frontend / docker compose / agent eval — n/a for this PR.

### Findings (inline — each is a merge blocker per team policy)
For the `/auto-implement-issue --continue-pr` punch list:

- 🛑 **B1** — `events.py:138-139` — `principal_name` / `target_name` are nullable but not given `None` defaults → required at construction. Required change: add `= None` defaults (CodeRabbit's diff); strengthen `test_optional_fields_accept_none` to also cover omission.
- ⚠️ **M1** — `events.py:108-130` (docstring at the class) — the `frozen=True` immutability claim overstates what pydantic v2 actually guarantees for nested mutable values. Required change: tighten the docstring to acknowledge `event.payload["k"] = "v"` is NOT blocked and that the redaction contract is enforced upstream by `redact_payload()`, not by model immutability. Adopt the lighter docstring fix rather than CodeRabbit's `MappingProxyType` migration (see "Adjacent improvements" for why).
- ⚠️ **M2** — `tests/test_broadcast_events.py:95-104` — `test_payload_default_is_empty_dict` constructs both events via `_make_event()` which always supplies a payload, so the model's `default_factory=dict` path is never traversed. Required change: build the events without a payload kwarg + add `assert a.payload == {}` so a future regression that swaps `default_factory=dict` for `default={}` is caught.

### Adjacent improvements (non-blocking, not posted inline)
Items the author may pick up in a future PR. None block merge.

- CodeRabbit suggested the `Mapping[str, Any]` + `MappingProxyType` migration for `payload` to make the field genuinely immutable. Heavy-lift change (pydantic v2's `Mapping` handling has quirks around JSON serialisation; needs a `model_validator(mode='after')` to wrap on construction). For T2's blast radius (T3 publisher hands off to XADD; T4/T6 consumers always deserialise from fresh JSON) the practical mutation window is zero — a docstring fix (M1) suffices. Worth tracking for v0.2.next if the broadcast pipeline grows transformation stages.
- T3 (#309) integration test should read back from the Valkey stream and assert forbidden keys (`path`, `key`, `filter`) are absent for the sensitive classes — this pins the redaction contract end-to-end, not just at the redactor function level. The PR's own docstring at `events.py:127-129` calls this out as deferred to T3 scope — correct.

### AI reviewer (CodeRabbit) cross-check
- Adopted: 3 (B1 → CR optional-fields; M1 → CR frozen-immutability, downgraded from heavy-lift to docstring; M2 → CR default-factory).
- Disputed: 0.
- Out of scope: 0.

### Docs
- [x] `docs/codebase/backend.md` — broadcast substrate section extended with T2 paragraph. Verified no .NET analogies. T1 paragraph correctly updated to reference T3-T6 only.

### What worked well (review-body only)
- **`classify_op` precedence pins are well-written.** The match-order matters for policy (credential allowlist beats `.list` suffix; `audit.` prefix beats both `.list` and `.query` suffixes); the corresponding tests pin both invariants explicitly with docstrings that name the bug class they catch. Future "let me just refactor this to a single regex" temptations will fail those tests loudly.
- **`_maybe_row_count` defensive coercion** — handling stringified counts from a JSON round-trip with int-coercion + fail-closed-to-None is exactly right for an audit-query response field that may travel through serialisation layers before reaching the redactor. The test `test_audit_query_handles_stringified_row_count` is the proof.
- **Module docstring's "Why a classifier rather than per-op annotation"** — names the alternative honestly, makes the trade-off explicit, and ties it back to decision #3. Future "shouldn't we just annotate each op?" debates have a written rationale to push back against.

File: backend/src/meho_backplane/broadcast/events.py (hunk 1 above)
Comment (zdamir) [reply backend/src/meho_backplane/broadcast/events.py]: Fixed in `e7e8fc2` — added `= None` defaults to `principal_name` / `target_name`. Strengthened the test surface with `test_optional_fields_omittable_at_construction` that exercises the omission path (the prior `test_optional_fields_accept_none` only covered the explicit-`None` path). Verified locally: `BroadcastEvent(...)` without those fields now succeeds with both = None.

File: backend/src/meho_backplane/broadcast/events.py (hunk 2 above)
Comment (zdamir) [reply backend/src/meho_backplane/broadcast/events.py:139]: Fixed in `e7e8fc2` — tightened the class docstring: `frozen=True` is now correctly described as faux-immutability (blocks attribute reassignment but NOT in-place mutation of nested mutable values like `payload`). Explicit statement that the PII contract is enforced upstream by `redact_payload()` before construction, not by the model. Noted the `Mapping[str, Any]` + `MappingProxyType` migration as tracked out-of-scope for T2 (heavy-lift; T2's mutation window is microseconds; revisit if the broadcast pipeline grows transformation stages).

File: backend/tests/test_broadcast_events.py (hunk 3 above)
Comment (zdamir) [reply backend/tests/test_broadcast_events.py:142]: Fixed in `e7e8fc2` — rewrote the test to `model_dump().pop('payload')` + reconstruct so `default_factory=dict` actually runs. Added `assert a.payload == {}` / `assert b.payload == {}` so a future regression that swaps `default_factory=dict` for `default={}` (the shared-mutable-default footgun) fails the test loudly.

**Comment by zdamir** [comment]
@coderabbitai review

**Review by zdamir — COMMENTED** [review]
> **Note:** Posting as `--comment` because GitHub blocks formal `APPROVE` on one's own PR. Formal verdict below is **APPROVE**; all three iter-1 findings are properly fixed in `e7e8fc2`.

## /review-pr-slim summary (iteration 2)

**Anchor task:** #308 — G6.1-T2 BroadcastEvent schema + classify_op + redact_payload (conservative PII defaults per decision #3)
**Surface area:** backend (Python)
**Best-practices packs audited:** python (pydantic v2 frozen-model semantics, optional vs nullable, test honesty)
**Verdict:** APPROVE
**Head SHA reviewed:** `e7e8fc2da2ac08693166f1a03158dc31f4a18d3e`
**AI reviewer (CodeRabbit):** the iter-2 re-trigger landed at 09:24:52 but CodeRabbit dedup-suppressed the re-review (its `09:09:48` "No actionable comments were generated in the recent batch" interim was already its iter-2 verdict). The original 3 findings auto-resolved (M1 docstring) / outdated (line shifts) / left-open (B1 thread on line 141 didn't auto-resolve, but the fix is verifiably correct — see below).

### Acceptance criteria
- [x] AC #1 — `BroadcastEvent` defined, frozen, all required fields enforced; optional fields now truly omittable — `TestBroadcastEventSchema` (+1 new test `test_optional_fields_omittable_at_construction`).
- [x] AC #2-5 — `classify_op` mapping unchanged from iter-1; all parametric cases still pass.
- [x] AC #6 — `redact_payload("credential_read", ...)` drops path/key — unchanged from iter-1.
- [x] AC #7 — `redact_payload("audit_query", ...)` drops filter — unchanged from iter-1.
- [x] AC #8 — `redact_payload("read", ...)` keeps params — unchanged from iter-1.
- [x] AC #9 — unit tests now actually exercise the default_factory + omission paths (M2 fix). All edge cases (empty params, nested objects, stringified row_count, omittable optional fields, default factory) covered.
- [x] AC #10 — CI green: DCO ✅, SonarCloud Quality Gate ✅, Python ruff+mypy+pytest ✅, Semgrep SAST/OSS ✅, Go ✅, Helm ✅, image build ✅, license checks ✅, CodeRabbit ✅.
- [ ] AC #11 — Initiative #228 body checklist for T2 — post-merge action.

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #308 (T2) | open | `Closes #308` | ✅ | ✅ full | close on merge, comment |
| #228 (Initiative G6.1) | open | parent of #308 | — | ⚠️ partial (T2 of 6) | tick T2 checkbox post-merge |
| #309 (T3 publish-on-write) | open | downstream consumer | — | — | unblocked once #308 lands |

### Predicted vs. actual diff (iter-2 delta from iter-1)
- ✅ `backend/src/meho_backplane/broadcast/events.py` — `principal_name` / `target_name` now have `= None` defaults (B1) + class docstring tightened to reflect actual `frozen=True` semantics (M1).
- ✅ `backend/tests/test_broadcast_events.py` — new `test_optional_fields_omittable_at_construction` exercises the omission path (B1); `test_payload_default_is_empty_dict` rewritten to actually traverse `Field(default_factory=dict)` (M2).
- ✅ No unexpected changes; iter-2 scope matches the punch list exactly.

### Integration verification
- [x] `ruff check src tests` — clean
- [x] `ruff format --check src tests` — clean (102 files)
- [x] `mypy src` strict — 53 source files, no issues
- [x] `pytest tests/test_broadcast_events.py` — **41 passed** (+1 over iter-1 baseline: the new omittable-fields test)
- [x] `pytest --ignore=tests/integration` — **514 passed** (+1 vs iter-1, +41 vs main pre-PR), 1 skipped, 1 deselected
- [x] **B1 standalone repro** — `BroadcastEvent(...)` without `principal_name` / `target_name` now succeeds with both = None (was: `ValidationError` in iter-1). The contract bug is verifiably closed.
- [⚠️ not run] migrations / docker compose / agent eval — n/a.

### Findings (inline)
None. All three iter-1 findings (B1 blocker + M1 / M2 majors) are fixed in `e7e8fc2`:

- **B1** — `events.py:151-152` carry `= None` defaults; new `test_optional_fields_omittable_at_construction` exercises the omission path; standalone Python repro confirms.
- **M1** — `events.py:108-138` (class docstring) tightened: `frozen=True` correctly described as faux-immutability in pydantic v2, explicit statement that `event.payload["k"] = "v"` is NOT blocked, PII contract location named as `redact_payload()` upstream of construction. `MappingProxyType` migration tracked as adjacent improvement.
- **M2** — `test_broadcast_events.py:119-141` (`test_payload_default_is_empty_dict`) rewritten via `model_dump().pop("payload")` + reconstruct so `Field(default_factory=dict)` actually runs. `assert == {}` added so a future regression that swaps `default_factory=dict` for `default={}` fails loudly.

### Adjacent improvements (non-blocking)
- The B1 review thread on `events.py:141` (CodeRabbit's original anchor at the post-fix line) is not marked `resolved` in the GraphQL `reviewThreads` view. CodeRabbit's auto-resolution heuristic looks for an exact pattern match against its diff suggestion; my fix landed the same `= None` defaults but added explanatory comment lines above (lines 147-150), which trips CR's pattern matcher. The thread should be marked resolved manually before merge to satisfy the "all comments resolved" team policy — it's a UI click, not a code action. The M2 thread auto-resolved; M1's thread was marked outdated when line numbers shifted.
- `Mapping[str, Any]` + `types.MappingProxyType` migration for genuine payload immutability remains a v0.2.next candidate. The M1 docstring now names it explicitly as out-of-scope for T2.

### Docs
- [x] `docs/codebase/backend.md` — updated in the original PR; iter-2 didn't touch.

### What worked well (review-body only)
- **B1 fix is more than a one-line diff** — the author added explanatory comment lines (147-150) above the now-defaulted fields naming the reason omission has to be cheap. Future readers see the contract intent inline, not just the default value.
- **M2 test rewrite includes a docstring naming the specific footgun it catches** (`default={}` shared-mutable-default). A future regression that re-introduces it now has a named test to fail — much sharper than the prior shape that asserted only `is not`.

## Post-merge actions (do after merge)

### Auto-close (should happen automatically)
- #308 — verify it actually closed (same-repo `Closes #308`, default-branch base, no squash-strip risk).

### Project-board status bumps
- #308 → `Done`
- #228 — comment "T2 landed; T3 (#309 publish-on-write) is the next dep"; tick T2 in Initiative body.

### Manual cleanup
- Mark the B1 CodeRabbit thread on `events.py:141` as resolved (auto-resolution didn't fire due to CR's pattern matcher; the fix is correct).

### Completion comments to post on close
- #308: "Closed by #346. `BroadcastEvent` + `classify_op` + `redact_payload` with decision-#3 conservative PII defaults; 41 unit tests + B1 omittable-fields contract pinned."

**Review by ikaric — APPROVED** [review]
